src/aiotaskq/interfaces.py

```diff
@@ -0,0 +1,116 @@
+"""
+Define all interfaces for the library.
+
+Interfaces are mainly typing.Protocol classes, but may also include
+other declarative classes like enums or Types.
+"""
+
+import enum
+import typing as t
+
+
+Message = t.Union[str, bytes]
+
+
+class PollResponse(t.TypedDict):
+    """Define the dictionary returned from a pubsub."""
+
+    type: str
+    data: Message
+    pattern: t.Optional[str]
+    channel: bytes
+
+
+class IProcess(t.Protocol):
+    """
+    Define the interface for a process used in the library.
+
+    It's more or less the same as the `multiprocessing.Process` except this
+    one only has attributes that are necessary for the library, and also has
+    slightly different typing e.g. pid in our case is always an `int`, whereas
+    the one from `multiprocessing.Process` is `Optional[int]`. This way we're
+    not limited to `multiprocessing.Process` and may switch to another implementation
+    if needed.
+    """
+
+    @property
+    def pid(self) -> t.Optional[int]:
+        """Return the process id (pid)."""
+
+    def start(self):
+        """Start running the process."""
+
+    def terminate(self):
+        """Send TERM signal to the process."""
+
+
+class ConcurrencyType(str, enum.Enum):
+    """Define supported concurrency types."""
+
+    MULTIPROCESSING = "multiprocessing"
+
+
+class IConcurrencyManager(t.Protocol):
+    """
+    Define the interface of a concurrency manager.
+
+    It should be able to start x number of processes given & terminate them.
+    """
+
+    concurrency: int
+    processes: dict[int, IProcess]
+
+    def __init__(self, concurrency: int) -> None:
+        """Initialize the concurrency manager."""
+
+    def start(self, func: t.Callable, *args: t.ParamSpecArgs) -> None:
+        """Start each process under management."""
+
+    def terminate(self) -> None:
+        """Terminate each process under management."""
+
+
+class IPubSub(t.Protocol):
+    def __init__(self, url: str, poll_interval_s: float, *args, **kwargs):
+        """Initialize the pubsub class."""
+
+    async def __aenter__(self) -> "IPubSub":
+        """Instantiate/start resources when entering the async context."""
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
```

Comment:
I felt like anything `__xxx__` should not be part of the interface?

Reply:
I disagree. To me, `__xxx__` is one of the important interface in python. Take `__init__` and `__len__` for example, they are very important since they let the users know that yes this class A supports being constructed as such and a such, and an instance of A can be called `len()` on it to get info about its size. All these without the users know exactly how it was constructed or how the size was retrieved. 

In this case, we're telling the user of this code that if they want to implement a pubsub it must support asynchronous context manager (because `__aenter__` and `__aexit__`) and that it should be constructed with a url, poll_interval_s and some other optional *args, **kwargs that they want depending on however they implement it.